File: DSA/Stack/valid_parenthesis.cpp
#include<bits/stdc++.h>
using namespace std;

bool isValidParenthesis(string str)
{
    stack<char>st;

    for(int i = 0;i < str.size();i++)
    {
        if(str[i] == '(' ||  str[i] == '{' || str[i] == '[')
            st.push(str[i]);
        
        else {
            if(st.empty())
                return false;
            else {
                if(str[i] == ')' && st.top() != '(')
                    return false;
                if(str[i] == '}' && st.top() != '{')
                    return false;
                if(str[i] == ']' && st.top() != '[')
                    return false;

                st.pop();
            }
        }
    }
    return st.empty();
}

int main()
{
    string str = "{{}}";
    if(isValidParenthesis(str) == true)
        cout << "Valid Parenthesis";
    else
        cout << "Not Valid Parenthesis";
    
    return 0;
}   

File: Python/Stack/valid_parenthesis.py
def is_valid_parenthesis(s):
    stack = []

    for char in s:
        if char in "({[":
            stack.append(char)
        else:
            if not stack:
                return False
            top = stack.pop()
            if char == ')' and top != '(':
                return False
            if char == '}' and top != '{':
                return False
            if char == ']' and top != '[':
                return False

    return not stack

def main():
    s = "{{}}"
    if is_valid_parenthesis(s):
        print("Valid Parenthesis")
    else:
        print("Not Valid Parenthesis")

if __name__ == "__main__":
    main()
